packages/core/src/directory.ts

```diff
@@ -0,0 +1,263 @@
+import { Org, OrgID, OrgMember, OrgMemberStatus, Group } from "./org";
+import { Server } from "./server";
+import { getIdFromEmail } from "./util";
+import { Auth } from "./auth";
+
+export interface DirectoryUser {
+    externalId?: string;
+    email?: string;
+    name?: string;
+    active?: boolean;
+}
+
+export interface DirectoryGroup {
+    externalId?: string;
+    name: string;
+    members?: DirectoryUser[];
+}
+
+export interface DirectoryGroupChanges {
+    newName?: string;
+    memberIdsToAdd?: string[];
+    memberIdsToRemove?: string[];
+}
+
+export interface DirectorySubscriber {
+    userCreated(user: DirectoryUser, orgId: OrgID): Promise<OrgMember | void>;
+    userUpdated(user: DirectoryUser, orgId: OrgID, userId: string): Promise<OrgMember | void>;
+    userDeleted(user: DirectoryUser, orgId: OrgID, userId: string): Promise<void>;
+
+    groupCreated(group: DirectoryGroup, orgId: OrgID): Promise<Group | void>;
+    groupUpdated(groupChanges: DirectoryGroupChanges, orgId: OrgID, groupId: string): Promise<Group | void>;
```

Comment:
I think I prefer `groupUpdated` to have the same signature as the other functions. Changes can (and probably should) be inferred by the directory subscriber by comparing the existing group with the new one.

Reply:
I actually would prefer `userUpdated` to have the same signature as this, because a `PATCH` should always have changes and not the full object. Also, for example, arrays can be very costly to "infer" when something has been removed, for example, especially after a few thousands of items. When a member is added, you need to fetch all previous members to build the "updated" version, and when it's removed, you need to fetch everything and remove the right item as well.

Reply:
If you look at my changes, you'll see that there is nothing to infer on the `DirectorySubscriber` side. We simply mirror whatever we get from the directory provider. Updating and inferring changes should happen on the `ScimServer` side of things.

Reply:
> When a member is added, you need to fetch all previous members to build the "updated" version, and when it's removed, you need to fetch everything and remove the right item as well.

I agree that it's less efficient, but I don't think the difference big enough to make a difference. Simply rebuilding the member lists on the other hand is much more robust and guarantees that everything is always in sync, even if an intermediate update fails.

Reply:
You can also see that the new code is much simpler and therefore less error-prone: https://github.com/padloc/padloc/pull/429/commits/00b33e801ee964dafbb84b96f49bf30a0d5e1bcc#diff-e48f537ea43df65f2f733b68e321624fd34237257fab618884ade5c6a77ef438L181-R178